File: lib/domain/model/heart_models.dart
import 'package:freezed_annotation/freezed_annotation.dart';

import '../../core/time/local_date.dart';

part 'heart_models.freezed.dart';

@freezed
abstract class HeartRateSample with _$HeartRateSample {
  const factory HeartRateSample({
    required DateTime time,
    required int beatsPerMinute,
    required String source,
  }) = _HeartRateSample;
}

@freezed
abstract class RestingHeartRateSample with _$RestingHeartRateSample {
  const factory RestingHeartRateSample({
    required DateTime time,
    required int beatsPerMinute,
    required String source,
  }) = _RestingHeartRateSample;
}

@freezed
abstract class HrvSample with _$HrvSample {
  const factory HrvSample({
    required DateTime time,
    required double rmssdMs,
    required String source,
  }) = _HrvSample;
}

@freezed
abstract class HeartRateSummary with _$HeartRateSummary {
  const factory HeartRateSummary({
    required LocalDate date,
    required int avgBpm,
    required int minBpm,
    required int maxBpm,
  }) = _HeartRateSummary;
}

@freezed
abstract class DailyRestingHR with _$DailyRestingHR {
  const factory DailyRestingHR({
    required LocalDate date,
    required int bpm,
  }) = _DailyRestingHR;
}

@freezed
abstract class DailyHrv with _$DailyHrv {
  const factory DailyHrv({
    required LocalDate date,
    required double rmssdMs,
  }) = _DailyHrv;
}

File: app/src/main/kotlin/tech/mmarca/openvitals/domain/model/HeartModels.kt
package tech.mmarca.openvitals.domain.model

import java.time.Instant
import java.time.LocalDate
import kotlin.math.roundToLong
import tech.mmarca.openvitals.core.stats.timeBucketedAverageOrNull

data class HeartRateSample(
    val time: Instant,
    val beatsPerMinute: Long,
    val source: String,
)

data class RestingHeartRateSample(
    val time: Instant,
    val beatsPerMinute: Long,
    val source: String,
)

/** A day's resting rate: the minute-bucketed mean of its samples, rounded. */
fun List<RestingHeartRateSample>.dayAverageBpm(): Long? =
    timeBucketedAverageOrNull(time = { it.time }, value = { it.beatsPerMinute.toDouble() })?.roundToLong()

data class HrvSample(
    val time: Instant,
    val rmssdMs: Double,
    val source: String,
    val id: String = "",
    val isOpenVitalsEntry: Boolean = false,
)

data class HeartRateSummary(
    val date: LocalDate,
    val avgBpm: Long,
    val minBpm: Long,
    val maxBpm: Long,
)

/**
 * One local day of hourly heart-rate aggregates. Inside an hour Health Connect
 * weights by sample, so [summary]'s average leans toward 1 Hz workouts.
 * [signature] changes when any hour's count, average, low or high does.
 */
data class HeartRateDayAggregate(
    val summary: HeartRateSummary,
    val sampleCount: Long,
    val signature: String,
)

data class DailyRestingHR(
    val date: LocalDate,
    val bpm: Long,
)

data class DailyHrv(
    val date: LocalDate,
    val rmssdMs: Double,
)
